devito/finite_differences/tools.py

```diff
@@ -0,0 +1,211 @@
+    """
+    pass
+
+
+direct = Transpose('direct', 1)
+transpose = Transpose('transpose', -1)
+
+
+class Side(Tag):
+    """
+    Class encapsulating the side of the shift for derivatives.
+    """
+
+    def adjoint(self, matvec):
+        if matvec == direct:
+            return self
+        else:
+            if self == centered:
+                return centered
+            elif self == right:
+                return left
+            elif self == left:
+                return right
+            else:
+                raise ValueError("Unsupported side value")
+
+
+left = Side('left', -1)
+right = Side('right', 1)
+centered = Side('centered', 0)
+
+
+def check_input(func):
+    @wraps(func)
+    def wrapper(expr, *args, **kwargs):
+        if expr.is_Number:
+            return S.Zero
+        elif not isinstance(expr, Differentiable):
+            raise ValueError("`%s` must be of type Differentiable (found `%s`)"
+                             % (expr, type(expr)))
+        else:
+            return func(expr, *args, **kwargs)
+    return wrapper
+
+
+def check_symbolic(func):
+    @wraps(func)
+    def wrapper(expr, *args, **kwargs):
+        if expr._uses_symbolic_coefficients:
+            expr_dict = expr.as_coefficients_dict()
+            if any(len(expr_dict) > 1 for item in expr_dict):
+                raise NotImplementedError("Applying the chain rule to functions "
+                                          "with symbolic coefficients is not currently "
+                                          "supported")
+        kwargs['symbolic'] = expr._uses_symbolic_coefficients
+        return func(expr, *args, **kwargs)
+    return wrapper
+
+
+def dim_with_order(dims, orders):
+    """
+    Create all possible derivative order for each dims
+    for example dim_with_order((x, y), 1) outputs:
+    [(1, 0), (0, 1), (1, 1)]
+    """
+    ndim = len(dims)
+    max_order = np.max(orders)
+    # Get all combinations and remove (0, 0 ,0)
+    all_comb = tuple(product(range(max_order+1), repeat=ndim))[1:]
+    # Only keep the one with each dimension maximum order
+    all_comb = [c for c in all_comb if all(c[k] <= orders[k] for k in range(ndim))]
+    return all_comb
+
+
+def deriv_name(dims, orders):
+    name = []
+    for d, o in zip(dims, orders):
+        name_dim = 't' if d.is_Time else d.root.name
+        name.append('d%s%s' % (name_dim, o) if o > 1 else 'd%s' % name_dim)
+
+    return ''.join(name)
+
+
+def generate_fd_shortcuts(function):
+    """Create all legal finite-difference derivatives for the given Function."""
+    dimensions = function.indices
+    s_fd_order = function.space_order
+    t_fd_order = function.time_order if (function.is_TimeFunction or
+                                         function.is_SparseTimeFunction) else 0
+    orders = tuple(t_fd_order if i.is_Time else s_fd_order for i in dimensions)
+
+    from devito.finite_differences.derivative import Derivative
```

Comment:
never local imports unless it's a pathological case

Reply:
You did that one so gonna assume pathological

Reply:
oki